freqtrade/strategy/hyper.py

```diff
@@ -0,0 +1,93 @@
+"""
+IHyperStrategy interface, hyperoptable Parameter class.
+This module defines a base class for auto-hyperoptable strategies.
+"""
+from abc import ABC
+from typing import Union, List, Iterator, Tuple
+
+from skopt.space import Integer, Real, Categorical
+
+from freqtrade.strategy.interface import IStrategy
+
+
+class Parameter(object):
+    """
+    Defines a parameter that can be optimized by hyperopt.
+    """
+    default: Union[int, float, str, bool]
+    space: List[Union[int, float, str, bool]]
+    category: str
+
+    def __init__(self, *, space: List[Union[int, float, str, bool]], default: Union[int, float, str, bool] = None,
+                 category: str = None, **kwargs):
+        """
+        Initialize hyperopt-optimizable parameter.
+        :param space: Optimization space. [min, max] for ints and floats or a list of strings for categorial parameters.
+        :param default: A default value. Required for ints and floats, optional for categorial parameters (first item
+         from the space will be used). Type of default value determines skopt space used for optimization.
+        :param category: A parameter category. Can be 'buy' or 'sell'. This parameter is optional if parameter field
+         name is prefixed with 'buy_' or 'sell_'.
+        :param kwargs: Extra parameters to skopt.space.(Integer|Real|Categorical).
+        """
+        assert 'name' not in kwargs, 'Name is determined by parameter field name and can not be specified manually.'
+        self.value = default
+        self.space = space
+        self.category = category
+        self._space_params = kwargs
+        if default is None:
+            assert len(space) > 0
+            self.value = space[0]
```

Comment:
I think i'd rather have "default" a required parameter (unless there's a good reason not to).

For categorical spaces using the first value might work ... but for an integer space, it's kinda odd ...

`Parameter(space=[-300, +300])` would pick -300, which is the negative extreme (and therefore rather unlikely to happen).

An alternative could be to use a "sensible default" - so a value in the middle in case of a `IntParameter` (see other comment), the first for the categorical ... but it's just an idea...

Reply:
-300 would not be picked for float/int as default must be set, but i do agree this stinks.`(Int|Float|Categorial)Parameter` would solve it cleanly, so one more reason to go that route.